### PR TITLE
feat(gRPC): implement BidiWriteObject

### LINKED ISSUE #568: Implement bidirectional writes in GRPC

### PR DESCRIPTION
**Pull request by cojenco**
Fixes #568

### CODE REVIEW

File: gcs/upload.py
Comment (coryan) [inline gcs/upload.py:286]: optional: this is a rather long function, consider breaking it down, maybe split up the function to process the first message?

File: tests/test_upload.py
Comment (coryan) [inline tests/test_upload.py:697]: Hmm... why?  Consider this scenario:

- The client is trying to resume an upload
- The client sends a QueryResumableUpload() RPC and gets a result of saying that the upload is not completed.
- Some other client (or a delayed message in the service internal processes) completes the upload
- The client tries to resume the upload by sending the same data as before, and finalizing.

I would think that should succeed with the previously uploaded object.

File: tests/test_upload.py
Comment (coryan) [inline tests/test_upload.py:990]: Consider a test for "invalid data field":

https://app.codecov.io/gh/googleapis/storage-testbench/pull/586?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=googleapis#83ee5f7e84364b0abad9908f111e7409-R335

**Review by coryan — COMMENTED** [review]
A couple of optional things and a question.

File: tests/test_upload.py
Comment (coryan) [reply tests/test_upload.py]: I see this is fixed.  FWIW, don't feel that you need to wait for another review from me.

File: tests/test_upload.py
Comment (cojenco) [reply tests/test_upload.py:992]: On second thought, we probably don't need this invalid field check after all. I think in most scenarios we can rely on protobuf message checking - for a oneof group, checks if any field inside is set and raises ValueError for anything not defined in the message descriptor.

File: gcs/upload.py
Comment (cojenco) [reply gcs/upload.py:286]: Agreed, I'm still trying to see how this can be split up without affecting the generator

File: gcs/upload.py
Comment (coryan) [inline gcs/upload.py:354]: This is a complicated way to write `if True` since `should_checkpoint_session` is always `True`?

Consider rewriting the code to just always checkpoint and then leave a TODO to implement behavior that more closely follows what the service does.

File: gcs/upload.py
Comment (coryan) [inline gcs/upload.py:291]: Just an idea (no action needed): maybe we should let the client pick the behavior (via headers like we do for retry testing). That way can write tests for different scenarios (e.g. when the data is not flushed and we disconnect, or when the data is flushed before the client expects it to).

File: gcs/upload.py
Comment (coryan) [inline gcs/upload.py:368]: Checking this after doing `upload.media += content` seems like a bug.

Seems like this check should be around line 350 or so?

File: gcs/upload.py
Comment (coryan) [inline gcs/upload.py:363]: I think if `request.finish_write == True and request.state_lookup == True` you do not want to yield just the persisted bytes?

**Review by coryan — CHANGES_REQUESTED** [review]

File: gcs/upload.py
Comment (cojenco) [reply gcs/upload.py]: haha yes it's quite the illusion xD 
Updated with comments and linked to a TODO issue

File: gcs/upload.py
Comment (cojenco) [reply gcs/upload.py:371]: This is to handle cases with missing BidiWriteObjectRequest. I probably went overboard, and there's also a test case for this `gcs.upload.Upload.process_bidi_write_object_grpc(db, [], context)`
https://github.com/googleapis/storage-testbench/blob/65105b6fdf5efcf590326bac65650076b3677824/tests/test_upload.py#L646-L653

File: gcs/upload.py
Comment (cojenco) [reply gcs/upload.py]: Good catch. I've updated this part of the logic so a response with persisted_size is only yielded for **incomplete uploads**. Definitely not elegant... my thought process here is for completed uploads when `finish_write = True` to be finalized out of the loop (store full object checksums, create new object, insert to database, and yield final response with resource metadata)... thoughts?

File: gcs/upload.py
Comment (coryan) [reply gcs/upload.py:371]: Right, but if `upload` is None then line 359 would have raised an exception before you got here, right?   You probably want to detect the problem before the first use of `upload`.  Did I miss something?  If so, please just merge, otherwise please reorder the test to prevent ~crashes~ exceptions.

File: gcs/upload.py
Comment (coryan) [reply gcs/upload.py]: SGTM

**Review by coryan — APPROVED** [review]

File: gcs/upload.py
Comment (cojenco) [reply gcs/upload.py:371]: Inside the request loop on L307, we're checking if `upload` is None before the first use. 

This extra check is outside the request loop. It covers the rare case of missing requests, and prevents crashes in the attempt to finalize an upload

**Pull request merged**
